content/advent-2018/go-arrow.md

```diff
@@ -0,0 +1,531 @@
+
+The code below creates a pair of records `rec1` and `rec2` (each record can be seen as a set of rows in a database.)
+Each record contains two columns, the first one containing `int32` values and the second one `float64` values.
+Once a few entries have been added to each of the two records, a `Table` reader is created from these in-memory records.
+The chunk size for this `Table` reader is set to `5`: each column will hold at most 5 entries during the iteration over the in-memory `Table`.
+
+```go
+package main
+
+import (
+	"fmt"
+
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/array"
+	"github.com/apache/arrow/go/arrow/memory"
+)
+
+func main() {
+	mem := memory.NewGoAllocator()
+
+	schema := arrow.NewSchema(
+		[]arrow.Field{
+			arrow.Field{Name: "f1-i32", Type: arrow.PrimitiveTypes.Int32},
+			arrow.Field{Name: "f2-f64", Type: arrow.PrimitiveTypes.Float64},
+		},
+		nil, // no metadata
+	)
+
+	b := array.NewRecordBuilder(mem, schema)
+	defer b.Release()
+
+	b.Field(0).(*array.Int32Builder).AppendValues(
+		[]int32{1, 2, 3, 4, 5, 6},
+		nil,
+	)
+	b.Field(0).(*array.Int32Builder).AppendValues(
+		[]int32{7, 8, 9, 10},
+		[]bool{true, true, false, true},
+	)
+	b.Field(1).(*array.Float64Builder).AppendValues(
+		[]float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 
+		nil,
+	)
+
+	rec1 := b.NewRecord()
+	defer rec1.Release()
+
+	b.Field(0).(*array.Int32Builder).AppendValues([]int32{
+		11, 12, 13, 14, 15, 16, 17, 18, 19, 20}, 
+		nil,
+	)
+	b.Field(1).(*array.Float64Builder).AppendValues([]float64{
+		11, 12, 13, 14, 15, 16, 17, 18, 19, 20}, 
+		nil,
+	)
+
+	rec2 := b.NewRecord()
+	defer rec2.Release()
+
+	tbl := array.NewTableFromRecords(schema, []array.Record{rec1, rec2})
+	defer tbl.Release()
+
+	tr := array.NewTableReader(tbl, 5)
+	defer tr.Release()
+
+	n := 0
+	for tr.Next() {
+		rec := tr.Record()
+		for i, col := range rec.Columns() {
+			fmt.Printf("rec[%d][%q]: %v\n", n, rec.ColumnName(i), col)
+		}
+		n++
+	}
+}
+```
+
+Executing the code above would result in:
+```
+$> go run ./table-reader.go
+rec[0]["f1-i32"]: [1 2 3 4 5]
+rec[0]["f2-f64"]: [1 2 3 4 5]
+rec[1]["f1-i32"]: [6 7 8 (null) 10]
+rec[1]["f2-f64"]: [6 7 8 9 10]
+rec[2]["f1-i32"]: [11 12 13 14 15]
+rec[2]["f2-f64"]: [11 12 13 14 15]
+rec[3]["f1-i32"]: [16 17 18 19 20]
+rec[3]["f2-f64"]: [16 17 18 19 20]
+```
+
+## CSV data
+
+Finally, to conclude with our quick whirlwind overview of what Go Arrow can provide now, let us consider the [arrow/csv](https://godoc.org/github.com/apache/arrow/go/arrow/csv) package.
+As mentioned earlier, many analysis pipelines start with ingesting [CSV](https://en.wikipedia.org/wiki/Comma-separated_values).
+The Go standard library already provides a package to decode and encode data using this comma-separated values (CSV) "format".
+But what the [encoding/csv](https://godoc.org/encoding/csv) package exposes is slices of `string`s, not values of a specific type.
+This is left to the user of that package.
```

Comment:
> But what the [encoding/csv] package exposes is slices of strings, not values of a specific type. This is left to the user of that package. 

Not sure I understand how are these two sentences connected. Maybe it would be better if you dropped the "This is left..." and went directly to [arrow/csv] part?

Reply:
done (explained what exactly was left to the user of `encoding/csv`.)